corehq/motech/openmrs/openmrs_config.py

```diff
@@ -149,3 +155,103 @@ class OpenmrsConfig(DocumentSchema):
+    """
+    Used for updating _property_map to map case properties to OpenMRS
+    patient property-, attribute- and concept values.
+
+    i.e. Allows us to answer the question, "If we know the case property how
+    do we find the OpenMRS value?"
+
+    :param jsonpath: The path to a value in an OpenMRS patient JSON object
+    :param value_source: A case_config ValueSource instance
+    :return: A single-item dictionary with the name of the case
+             property as key, and a JsonpathValuemap as value. If
+             value_source is a constant, then there is no corresponding
+             case property, so the function returns an empty dictionary
+    """
+    if value_source['doc_type'] == 'ConstantString':
+        return {}
+    if value_source['doc_type'] == 'CaseProperty':
+        return {value_source['case_property']: JsonpathValuemap(jsonpath, {})}
+    if value_source['doc_type'] == 'CasePropertyMap':
+        value_map = {v: k for k, v in value_source['value_map'].items()}
+        return {value_source['case_property']: JsonpathValuemap(jsonpath, value_map)}
+    raise ValueError(
+        '"{}" is not a recognised ValueSource for setting OpenMRS patient values from CommCare case properties. '
+        'Please check your OpenMRS case config.'.format(value_source['doc_type'])
+    )
+
+
+def get_property_map(case_config):
+    """
+    Returns a map of OpenMRS patient properties and attributes to case
+    properties.
+    """
+    property_map = {}
+
+    for person_prop, value_source in case_config['person_properties'].items():
+        jsonpath = parse_jsonpath('person.' + person_prop)
+        property_map.update(get_caseproperty_jsonpathvaluemap(jsonpath, value_source))
+
+    for attr_uuid, value_source in case_config['person_attributes'].items():
+        # jsonpath_rw offers programmatic JSONPath expressions. For details on how to create JSONPath
+        # expressions programmatically see the
+        # `jsonpath_rw documentation <https://github.com/kennknowles/python-jsonpath-rw#programmatic-jsonpath>`__
+        #
+        # The `Where` JSONPath expression "*jsonpath1* `where` *jsonpath2*" returns nodes matching *jsonpath1*
+        # where a child matches *jsonpath2*. `Cmp` does a comparison in *jsonpath2*. It accepts a
+        # comparison operator and a value. The JSONPath expression below is the equivalent of::
+        #
+        #     (person.attributes[*] where attributeType.uuid eq attr_uuid).value
+        #
+        # This `for` loop will let us extract the person attribute values where their attribute type UUIDs
+        # match those configured in case_config['person_attributes']
+        jsonpath = Child(
+            Where(
+                Child(Child(Fields('person'), Fields('attributes')), Slice()),
+                Cmp(Child(Fields('attributeType'), Fields('uuid')), eq, attr_uuid)
+            ),
+            Fields('value')
+        )
```

Comment:
what's the benefit of writing it like this vs having jsonpath parse the string version?

Reply:
We wrote `Cmp`, but extending the jsonpath parser to support it was going to take more time. I figured it was better just to use it like this for now, and extend the parser as a nice-to-have later.